x/subkeys/internal/types/expiration.go

```diff
@@ -0,0 +1,107 @@
+package types
+
+import "time"
+
+// ExpiresAt is a point in time where something expires.
+// It may be *either* block time or block height
+type ExpiresAt struct {
+	Time   time.Time `json:"time" yaml:"time"`
+	Height int64     `json:"height" yaml:"height"`
+}
+
+// ExpiresAtTime creates an expiration at the given time
+func ExpiresAtTime(t time.Time) ExpiresAt {
+	return ExpiresAt{Time: t}
+}
+
+// ExpiresAtHeight creates an expiration at the given height
+func ExpiresAtHeight(h int64) ExpiresAt {
+	return ExpiresAt{Height: h}
+}
+
+// ValidateBasic performs basic sanity checks.
+// Note that empty expiration is allowed
+func (e ExpiresAt) ValidateBasic() error {
+	if !e.Time.IsZero() && e.Height != 0 {
+		return ErrInvalidPeriod("both time and height are set")
+	}
+	if e.Height < 0 {
+		return ErrInvalidPeriod("negative height")
+	}
+	return nil
+}
+
+// IsZero returns true for an uninitialized struct
+func (e ExpiresAt) IsZero() bool {
+	return e.Time.IsZero() && e.Height == 0
+}
+
+// IsExpired returns if the time or height is *equal to* or greater
+// than the defined expiration point. Note that it is expired upon
+// an exact match.
+//
+// Note a "zero" ExpiresAt is never expired
+func (e ExpiresAt) IsExpired(t time.Time, h int64) bool {
+	if !e.Time.IsZero() && !t.Before(e.Time) {
+		return true
+	}
+	return e.Height != 0 && h >= e.Height
+}
+
+// IsCompatible returns true iff the two use the same units.
+// If false, they cannot be added.
+func (e ExpiresAt) IsCompatible(p Period) bool {
+	if !e.Time.IsZero() {
+		return p.Clock > 0
+	}
+	return p.Block > 0
+}
+
+// Step will increase the expiration point by one period
+// It returns an error if the period is incompatible
+func (e *ExpiresAt) Step(p Period) error {
```

Comment:
How and why is this used? Expiration should always be compared against the current block time and/or height. Do we even need this?

Reply:
It's still unclear to me why we need to do any manual "incrimenting"? Can you not just compare against the height/time? Look at the vesting account logic where we also have periodic vesting. There is no manual stepping there.